pkg/services/object/get/container.go

```diff
@@ -15,71 +17,84 @@ func (exec *execCtx) executeOnContainer() {
 
 	exec.log.Debug("trying to execute in container...")
 
-	// initialize epoch number
-	epoch, err := exec.svc.currentEpochReceiver.currentEpoch()
+	nodeLists, primaryCounts, err := exec.svc.neoFS.GetNodesForObject(exec.address())
 	if err != nil {
 		exec.status = statusUndefined
 		exec.err = err
-		exec.log.Debug("could not get current epoch number", zap.Error(err))
+		exec.log.Debug("failed to list storage nodes for the object", zap.Error(err))
 		return
 	}
 
-	exec.processEpoch(epoch)
-}
-
-func (exec *execCtx) processEpoch(epoch uint64) bool {
-	exec.log.Debug("process epoch",
-		zap.Uint64("number", epoch),
-	)
-
-	traverser, ok := exec.generateTraverser(exec.address(), epoch)
-	if !ok {
-		return true
-	}
-
 	ctx, cancel := context.WithCancel(exec.context())
 	defer cancel()
 
 	exec.status = statusUndefined
 	mProcessedNodes := make(map[string]struct{})
-
-	for {
-		addrs := traverser.Next()
-		if len(addrs) == 0 {
-			exec.log.Debug("no more nodes, abort placement iteration")
-
-			return false
+	var endpoints, externalEndpoints network.AddressGroup
+	var j, jLim uint
+	primary := true
+
+	for i := 0; i < len(nodeLists); i++ { // do not use for-range!
```

Comment:
why? the comment does not explain, so it is useless (or even harmful)

Reply:
usually range is used for this, but here it wouldn't work

Reply:
only after reading it twice (to the end), i saw `i = -1`, why not mention it? it is a comment, it should help. any code can be changed, so comments should explain something, and if they are outdated, it should be easy to understand. not saying it is required here but it is a general rule i would say

Reply:
Tricky one. Can be simplified by inverting i/j loops: walk over the top ones from each list first and then go deeper till there are no more nodes to check. It can be more optimal in some cases (top nodes from different lists are still more likely to have the object), but maybe not for all of them (policies may have local nodes first).

The other way of course is just `for _, primary := range []bool{true, false}` and a regular range loop over `nodeLists` then. Less tricks, same meaning.

Reply:
> top nodes from different lists are still more likely to have the object

they are homogeneous in terms of policies

> `for _, primary := range []bool{true, false}`

i can make any style you think is readable. I'm fine with current. It can be simply divided into two loops. No tricks at all